### PR TITLE
server/lock_manager: Support reporting the full wait chain when deadlock is detected

### PR DESCRIPTION
**Pull request by MyonKeminta**
~~This PR is now reviewable. But before merging, I want to test if there's any performance regression and try to optimize it if necessary.~~

### What problem does this PR solve?

Problem Summary:

We are going to implement the feature [Lock View](https://github.com/pingcap/tidb/issues/24199) . A part of it is to support collecting information about recent deadlocks.

### What is changed and how it works?

What's Changed:  Two more fields `key` and `resource_group_tag` is added to deadlock detection request, deadlock detector's inner information, and deadlock error message. SQL Digest will be encoded into `resource_group_tag`. The `resource_group_tag` may also be used by another feature *Top SQL*, and they want to pass both SQL Digest and the plan to TiKV by the `resource_group_tag`. 

### Related changes

- Requires kvproto: https://github.com/pingcap/kvproto/pull/752

### Check List 

Tests 

- Unit test
- Manual test
  - Tested several simple workload with this PR: https://github.com/pingcap/tidb/pull/24380 , and compared with the merge base to master.
  - There are 1 TiDB node and 3 TiKV nodes, each has 16c cpu adn 90G mem (TiDB has 64G), nvme disk.
  - No observable performance regression is found, no observable difference is found in TiKV's metrics.
 
| With this PR | Workload | Threads | Table size | QPS | .95 lat |
|----|--|--|--|--|--|
| Y | oltp_read_write | 32  | 100,000 | 14455.2  | 54.83 |
| Y | oltp_read_write | 128 | 100,000 | 24456.78 | 127.81|
|   | oltp_read_write | 32  | 100,000 | 14522.35 | 54.83 |
|   | oltp_read_write | 128 | 100,000 | 24734.54 | 127.81|
| Y | oltp_write_only | 32  | 100,000 | 10184.40 | 31.37 |
| Y | oltp_write_only | 128 | 100,000 | 21315.33 | 80.03 |
|   | oltp_write_only | 32  | 100,000 | 10227.46 | 31.37 |
|   | oltp_write_only | 128 | 100,000 | 20600.07 | 82.96 |

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

* No release note

### CODE REVIEW

File: src/storage/txn/commands/acquire_pessimistic_lock.rs [hunk 1]
```diff
@@ -60,14 +60,17 @@ impl CommandExt for AcquirePessimisticLock {
     gen_lock!(keys: multiple(|x| &x.0));
 }
 
-fn extract_lock_from_result<T>(res: &StorageResult<T>) -> Lock {
+fn extract_lock_key_from_result<T>(res: &StorageResult<T>) -> (Lock, Vec<u8>) {
```
Comment (longfangsong) [inline src/storage/txn/commands/acquire_pessimistic_lock.rs:63]: Can we just return `info` for this fuction, and let the only user(` process_write`) do the rest part?

File: src/storage/txn/commands/mod.rs [hunk 2]
```diff
@@ -364,8 +364,8 @@ pub struct WriteResult {
     pub to_be_write: WriteData,
     pub rows: usize,
     pub pr: ProcessResult,
-    // (lock, is_first_lock, wait_timeout)
-    pub lock_info: Option<(lock_manager::Lock, bool, Option<WaitTimeout>)>,
+    // (lock, key, is_first_lock, wait_timeout)
+    pub lock_info: Option<(lock_manager::Lock, Vec<u8>, bool, Option<WaitTimeout>)>,
```
Comment (longfangsong) [inline src/storage/txn/commands/mod.rs:368]: This tuple is a little too long now, can we make it a struct?

**Review by longfangsong — APPROVED** [review]
Rest LGTM, but we should merge [kvproto#752](https://github.com/pingcap/kvproto/pull/752) and use kvproto on master first.

File: src/server/lock_manager/waiter_manager.rs [hunk 3]
```diff
@@ -529,9 +544,19 @@ impl WaiterManager {
...
+        waiter_ts: TimeStamp,
+        lock: Lock,
+        deadlock_key_hash: u64,
+        wait_chain: Vec<WaitForEntry>,
+    ) {
         if let Some(mut waiter) = self.wait_table.borrow_mut().remove_waiter(lock, waiter_ts) {
-            waiter.deadlock_with(deadlock_key_hash);
+            // The `wait_chain` doesn't include the final edge that caused the deadlock, which
+            // may reduce some deadlock RPC size. To pass it to the client, append this edge
+            // to the wait chain.
```
[the remark this answers is not in the document]
Comment (youjiali1995) [reply src/server/lock_manager/waiter_manager.rs:557]: You have appended it in the deadlock detector. It should be complete. I prefer to remove these comments.

**Review by youjiali1995 — COMMENTED** [review]
Rest LGTM.

File: src/server/lock_manager/waiter_manager.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply src/server/lock_manager/waiter_manager.rs:557]: Ah sorry. This comment is about the first implementation and later I changed it.

**Review by youjiali1995 — APPROVED** [review]

**Pull request merged**